src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/linalg/IterativeLinearSolverNDArray.java

```diff
@@ -0,0 +1,183 @@
+            z = r; /* by reference */
+        } else {
+            z = Nd4j.create(r.shape());
+        }
+        double rnorm = normFunc.apply(r);
+        if (rnorm <= rmax) {
+            return generateCGSignal(x, iter, rnorm, rmax, ExitStatus.SUCCESS_REL_TOL);
+        }
+
+        double rhoPrev = 0;
+        while (true) {
+            /* z_{i-1} = M r_{i-1} */
+            if (precond != null) {
+                z.assign(precond.operate(r));
+            }
+            /* rho_{i-1} = r_{i-1}^T z_{i-1} */
+            final double rhoNext = innerProductFunc.apply(r, z);
+            if (performDetailedChecks && rhoNext <= 0) {
+                throw new NonPositiveDefiniteOperatorException();
+            }
+            if (iter == 0) {
+                /* p_1 = z_0 */
+                p.assign(z);
+            } else {
+                /* p_{i} = z_{i-1}  + (\rho_{i-1}/\rho_{i-2}) p_{i-1} */
+                p.muli(rhoNext / rhoPrev).addi(z);
+            }
+            /* q_{i} = A p_{i} */
+            q.assign(linop.operate(p));
+            final double pq = innerProductFunc.apply(p, q);
+            if (performDetailedChecks && pq <= 0) {
+                throw new NonPositiveDefiniteOperatorException();
+            }
+            /* \alpha_{i} = \rho_{i-1}/(p_{i}^T q_{i}^T) */
+            final double alpha = rhoNext / pq;
+            /* x_{i} = x_{i-1}  + \alpha_{i} p_{i} */
+            x.addi(p.mul(alpha));
+            /* r_{i} = r_{i-1}  - \alpha_{i} q_{i} */
+            r.addi(q.mul(-alpha));
+            rhoPrev = rhoNext;
+            rnorm = normFunc.apply(r);
+            if (rnorm <= rmax) {
+                return generateCGSignal(x, iter, rnorm, rmax, ExitStatus.SUCCESS_REL_TOL);
+            }
+            if (rnorm <= absTol) {
+                return generateCGSignal(x, iter, rnorm, rmax, ExitStatus.SUCCESS_ABS_TOL);
+            }
+            iter += 1;
+            if (iter >= maxIters) {
+                return generateCGSignal(x, iter, rnorm, rmax, ExitStatus.FAIL_MAX_ITERS);
+            }
+        }
+    }
+
+    /**
+     * Generate an exit signal for {@link #cg(INDArray)}
+     *
+     * @param x solution
+     * @param iter number of iterations
+     * @param resNorm residual error norm
+     * @param resNormThreshold latest solver stopping criterion on residual error norm
+     * @param status the exist status of the solver
+     * @return a instance of {@link SubroutineSignal}
+     */
+    private static SubroutineSignal generateCGSignal(@Nonnull final INDArray x,
+                                                    final int iter,
+                                                    final double resNorm,
+                                                    final double resNormThreshold,
+                                                    @Nonnull final ExitStatus status) {
+        return SubroutineSignal.builder()
+                .put("x", x)
+                .put("iterations", iter)
```

Comment:
OK, all `SubroutineSignal` sets and gets are handled with an enum `StandardSubroutineSignals` now.